### DISCUSSION TITLE
Runtime pattern matching

### DISCUSSION THREAD

**Discussion by ziaulhasanhamim** [dotnet/csharplang#6441] [Language Ideas]
C# with pattern matching have come too far. But C# doesn't support runtime variables in a pattern? While many langues support that.
```C#
var txt = num switch
{
  0 => "Zero",
  _ => "NonZero"
}; // ok

var zero = 0;
var txt = num switch
{
  zero => "Zero",
  _ => "NonZero"
}; // error
```
In languages like Kotlin and go you do something like this

```kotlin
when(true) { 
  x.isOdd() -> print("x is odd") 
  y.isEven() -> print("y is even") 
  else -> print("x+y is odd") 
}
// they even have a shorthand for that which implies it is useful to have this
when { 
  x.isOdd() -> print("x is odd") 
  y.isEven() -> print("y is even") 
  else -> print("x+y is odd") 
}
```

**Comment by Richiban** [comment]
Much of this scenario is covered by [active patterns](https://github.com/dotnet/csharplang/issues/277).

**Reply by ziaulhasanhamim** [reply]
That's a whole different proposal containing a lot more things. And that is closed too

**Reply by HaloFour** [reply]
Here's the open proposals for active patterns:

https://github.com/dotnet/csharplang/issues/4131
https://github.com/dotnet/csharplang/issues/1047

It's something that the team has expressed interest in.

It sounds like your interest is more in removing the requirement that values used in patterns be constants.  That has been discussed before.  I'll see if I can dig up the discussion, it's probably buried in comments.  I think the concerns were mostly about exhaustiveness checking which would go right out the window but that there weren't any real technical reasons it couldn't work with variables.

**Reply by HaloFour** [reply]
Here's a thread of comments on the subject: https://github.com/dotnet/csharplang/issues/2850#issuecomment-623179988

As mentioned, many/most languages with pattern matching don't seem to allow matching against variables.

**Reply by ziaulhasanhamim** [reply]
But those languages which do support this thing aren't seeing any consequences either. Like in C# also it is still possible to do checks on variables using when but that's not very intuitive. I think the only reason most languages still don't support runtime variables in pattern is that C switch didn't support that.

**Reply by HaloFour** [reply]
>  I think the only reason most languages still don't support runtime variables in pattern is that C switch didn't support that.

I'm pretty sure that this is not the case.  Languages like Haskell/Scala/Rust don't take their inspiration from C.  I think it's more a paradigmic thing.  In a lot of languages matching against an identifier is a variable pattern, C# seems pretty unique in that the declaration requires additional syntax.

**Comment by Richiban** [comment]
I actually do like the idea of having a `switch` form that _doesn't_ have a receiver, sort of equivalent to an expression-if-else chain. Something like this:

```csharp
	bool cond1 = ...
	bool cond2 = ...

	var r = switch 
	{
		cond1 => "cond1 is true",
		cond2 => "cond2 is true",
		cond1 == cond2 => "cond1 and 2 are the same",
		_ => "Something else happened"
	}
```

which would be functionally equivalent to a switch with a fake receiver object:

```csharp
	var r = ((object)null) switch
	{
		_ when cond1 => "cond1 is true",
		_ when cond2 => "cond2 is true",
		_ when cond1 == cond2 => "cond1 and 2 are the same",
		_ => "Something else happened"
	}
```	

or the following if-else chain:

```csharp
        string r;

	if (cond1)
	{
		r = "cond1 is true";
	} else if (cond2)
	{
		r = "cond2 is true";
	} else if (cond1 == cond2)
	{
		r = "cond1 and 2 are the same";
	} else
	{
		r = "Something else happened";
	}
```

This is similar to your last Kotlin example.

**Reply by ziaulhasanhamim** [reply]
There is no reason to not like it. I have used it in Go. I have seen it used in Kotlin. And there is no disadvantages of adding it. As shown you can do the same in C# too with a bit more verbose.

**Reply by HaloFour** [reply]
IMO that's not really that worthwhile, nor is it much of a benefit over the conditional operator:

```cs
var r = cond1 ? "cond1 is true" :
        cond2 ? "cond2 is true" :
        cond1 == cond2 ? "cond1 and 2 are the same" :
        "Something else happened";
```

You can argue over the terseness but it has been idiomatic C# since 1.0 and, IMO, doesn't require much formatting to make very easy to read.

The benefit also is limited to a set of boolean expressions and doesn't help with anything slightly complicated like recursive patterns.

**Reply by Richiban** [reply]
Do you know, I'd actually never seen chained ternary operators like this... yep, that works just as well.

**Reply by ziaulhasanhamim** [reply]
Ternaries are no way readable specially when you have nested conditions. That is the modern languages have if expressions over ternaries. And when is the last time you seen someone using ternaries. I haven't seen. But when clause is very useful. Check the languages which support that. You can do a lot with that.

**Reply by ziaulhasanhamim** [reply]
For me I use switch over ternary even for  slightly complicated expressions rather than breaking ternary to multiple line.
```C#
var result = await usersService.EmailExists(user) switch
{
  true => // some logic,
  false => // some logic
}; 
````

**Reply by HaloFour** [reply]
Maybe, but that's just a question of syntactic preference.  I personally don't see the value in adding a form of `when` to the C# language because it's effectively the same thing.

I'd rather explore allowing variables in patterns in general, which, when combined with `switch` expressions, would achieve the same thing but also be substantially more powerful.

**Reply by ziaulhasanhamim** [reply]
Then all the pattern matching is effectively the same thing. You could everything before it was there. Pattern matching didn't a single which was not able to do be fore C#. Same goes for switch expression. There were no need for that then. Switch statement effectively does the same thing. Most importantly there was no need for arrow function in C#. Before that C# had delegate anonymous function which effectively does the same thing. 
```C#
var action = delegate() 
{ 
    Console.WriteLine("Hello world");
};
```
So everything is just syntactic sugar which make the code more readable and easier to write.

**Reply by HaloFour** [reply]
> Then all the pattern matching is effectively the same thing. You could everything before it was there. Pattern matching didn't a single which was not able to do be fore C#.

With _substantially_ more code, sure.  The value is in how much code you can reduce without decreasing the meaning of that code.  It's not just a matter of preferring one keyword or punctuation over another.  Recursive patterns can easily reduce dozens of lines of code into a single condition.

> Same goes for switch expression. There were no need for that then. Switch statement effectively does the same thing.

Except that it's not an expression, which forces where it can and can't be used.

> Most importantly there was no need for arrow function in C#. Before that C# had delegate anonymous function which effectively does the same thing.

It does, but again with substantially more code, which in simple scenarios eclipses the logic that is core to the lambda to begin with.  Anonymous delegates were designed in an era where the primary use case would be event handlers.  If the prevalence of functional programming was envisioned prior to that point it's extremely likely that anonymous delegates would have never happened.

> So everything is just syntactic sugar which make the code more readable and easier to write.

That doesn't make that sugar equal, especially when there is existing syntax that does quite literally the same thing and is no more or less verbose.

**Reply by ziaulhasanhamim** [reply]
Do you ever see that syntax used? No i don't see that. Because that's not convenient. Specially no one would use that there is nesting

**Reply by CyrusNajmabadi** [reply]
> Do you ever see that syntax used? No i don't see that. Because that's not convenient. Specially no one would use that there is nesting

I see nested ternaries all the time.

**Reply by HaloFour** [reply]
I use them in both C# and Java.  That's why I know to format them in the matter I illustrated as most of the complaints about them evaporate with a little whitespace.  😄 

It's quite convenient, but like any language feature it needs to be learned.

**Reply by ziaulhasanhamim** [reply]
Oh java🤣🤣. Now I got your point.

**Reply by ziaulhasanhamim** [reply]
Just leave the argument about ternary but that when has a lot use cases. Check some kotlin code.

**Reply by HaloFour** [reply]
Different languages have different features.  While this syntax is useful Kotlin and Go's general approach to "pattern matching" is incredibly limited, especially since they lack destructuring and recursive patterns, both of which are pretty critical to pattern matching in functional programming and both are supported in C# (and in preview in Java in two weeks).  Kotlin's answer to that is effectively active patterns, which are a powerful feature that I hope does come to C# soon.  I'm more interested in that space than providing another way to solve for evaluating multiple boolean expressions.

**Reply by RikkiGibson** [reply]
I think the suggestion for "switch without receiver" is clearer than the nested ternary because the `switch` and `{}` more clearly indicates when you are entering another level of nesting in a "when true" branch. Compare the following:
```cs
_ = switch
{
    a => switch
    {
        b => c
        _ => d
    }
    _ => e
}


_ = a
    ? b
        ? c
        : d
    : e
```

All that said, I don't think this is an especially urgent/compelling scenario.

**Reply by ziaulhasanhamim** [reply]
> All that said, I don't think this is an especially urgent/compelling scenario.

Agree on that. It's a good feature to have but not urgent.

There are times also where couldn't just used ternary instead of switch specially when we will have block bodied switch expressions. That is another thing I'm eagerly waiting for complex assignment and some multi-line code in switch expression.
